src/codegen.cpp

```diff
@@ -2842,8 +2858,10 @@ static jl_cgval_t emit_call(jl_expr_t *ex, jl_codectx_t *ctx)
                 else {
                     fval = emit_expr(args[0], ctx);
                 }
-                if (ctx->linfo->def) // root this li in case it gets deleted from the cache in `f`
-                    jl_add_linfo_root(ctx->linfo, (jl_value_t*)li);
+                // keep a reference to the called function's roots array in case
+                // the method is replaced (PR #14301)
+                if (ctx->linfo->def && li->def && li->def->roots)
+                    jl_add_linfo_root(ctx->linfo, (jl_value_t*)li->def->roots);
```

Comment:
another possible source of roots is `sparam_vals`, which `emit_sparam` assumes has a root (but could be any isbits)

additionally, lets just stop removing methods from the cache once they've been returned from `jl_get_specialization1`. this should be a really small number (since we actively discourage overwriting methods), avoids this issue, and is a going to be needed for 265 anyways.